Sparse CSR CUDA: Add block sparse support for torch.triangular_solve

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68711
* #68710
* __->__ #68709

This PR adds support for triangular solver with a block CSR matrix.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @ngimel

Differential Revision: [D33066067](https://our.internmc.facebook.com/intern/diff/D33066067)

IvanYashchuk committed:
Sparse CSR CUDA: Add block sparse support for torch.triangular_solve

This PR adds support for triangular solver with a block CSR matrix.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block sparse support for torch.triangular_solve"


This PR adds support for triangular solver with a block CSR matrix.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block sparse support for torch.triangular_solve"


This PR adds support for triangular solver with a block CSR matrix.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block sparse support for torch.triangular_solve"


This PR adds support for triangular solver with a block CSR matrix.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block sparse support for torch.triangular_solve"


This PR adds support for triangular solver with a block CSR matrix.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block sparse support for torch.triangular_solve"


This PR adds support for triangular solver with a block CSR matrix.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk commented:
ROCm failure is real. Strange that it builds locally and tests passed, need to verify that again. Maybe an older hipSPARSE version is used in CI.

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block sparse support for torch.triangular_solve"


This PR adds support for triangular solver with a block CSR matrix.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -109,6 +109,251 @@ void inline col_indices_and_values_resize_(const Tensor& input, int64_t nnz) {
+  auto ndim = X_->dim();
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(ndim == 2);
+  bool is_X_row_major = (X_strides[ndim - 1] == 1);
+  // int ldx = is_X_row_major ? cuda_int_cast(X_strides[ndim - 2], "ldx")
```
nit: Did you mean to remove this?

IvanYashchuk replied:
Yes, I need to clean up this and there is another commented code is left in a few places.

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block sparse support for torch.triangular_solve"


This PR adds support for triangular solver with a block CSR matrix.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33066067).

IvanYashchuk commented:
ROCm build fails https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.3.1-py3.6-build/17076/

IvanYashchuk commented:
It's legit. It was confusing to me because I tested previously locally and everything was building fine. The problem was that I tested with ROCm 4.5.0 and CI has older version 4.3.1. I will add skips for ROCm.

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block sparse support for torch.triangular_solve"


This PR adds support for triangular solver with a block CSR matrix.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

Differential Revision: [D33066067](https://our.internmc.facebook.com/intern/diff/D33066067)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33066067).
